examples/getting_started/web_search.ipynb

```diff
@@ -0,0 +1,255 @@
+      "execution_count": null,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "MAX_RESULTS = 10\n",
+        "\n",
+        "\n",
+        "class WebSearchAgent(BaseModel):\n",
+        "    history: list[ChatCompletionMessageParam] = []\n",
+        "    max_results: int = MAX_RESULTS\n",
+        "\n",
+        "    @openai.call(\"gpt-4o-mini\", stream=True, tools=[parse_webpage, web_search])\n",
+        "    @prompt_template(\"\"\"\n",
+        "        SYSTEM:\n",
+        "        Your task is to answer a user's query.\n",
+        "\n",
+        "        You have access to the following tools:\n",
+        "        - `web_search`: Search the web for information (limited to {self.max_results} results).\n",
+        "        - `parse_webpage`: Parse the content of a webpage.\n",
+        "\n",
+        "        When calling the `web_search` tool, the `body` is simply the body of the search\n",
+        "        result. You MUST then call the `parse_webpage` tool to get the actual content\n",
+        "        of the webpage. It is up to you to determine which search results to parse.\n",
+        "\n",
+        "        Once you have gathered all of the information you need, generate your response,\n",
+        "        which should strike the right balance between brevity and completeness. You answer\n",
+        "        to the user's query should be concise yet comprehensive.\n",
+        "\n",
+        "        MESSAGES: {self.history}\n",
+        "        USER: {query}\n",
+        "        \"\"\")\n",
+        "    def _stream(self, query: str): ...\n",
+        "\n",
+        "    def _step(self, query: str):\n",
+        "        stream = self._stream(query)\n",
+        "        tools_and_outputs = []\n",
+        "        for chunk, tool in stream:\n",
+        "            if tool:\n",
+        "                tools_and_outputs.append((tool, tool.call()))\n",
+        "            else:\n",
+        "                print(chunk.content, end=\"\", flush=True)\n",
+        "        if stream.user_message_param:\n",
+        "            self.history.append(stream.user_message_param)\n",
+        "        self.history.append(stream.message_param)\n",
+        "        if tools_and_outputs:\n",
+        "            self.history += stream.tool_message_params(tools_and_outputs)\n",
+        "            self._step(\"\")\n",
+        "\n",
+        "    def run(self):\n",
+        "        while True:\n",
+        "            query = input(\"(User): \")\n",
+        "            if query.lower() in [\"exit\", \"quit\"]:\n",
+        "                break\n",
+        "            print(\"(Assistant): \", end=\"\", flush=True)\n",
+        "            self._step(query)\n",
+        "            print(\"\")"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "Let's break down the key components of our WebSearchAgent class:\n",
```

Comment:
ok, I'm not sure I love the structure here. I like that you're breaking down the key components, but I think the structure here still follows the cookbook structure too closely.

Remember, the goal of these guides is to provide a more comprehensive introduction / getting started walkthrough. This gives us an opportunity to slowly build things in smaller chunks and explain each part, building on the last part until we have our final agent.

For example, we can first show how to implement a basic call, then we can show streaming the call, then we can show putting the streaming call inside the `BaseModel` agent and adding history, then we can show adding tools (which we could further break down if we wanted to into showing just the web search tool and then showing we need to add the parse webpage tool to get the actual content for better responses).

All along the way we should be printing and showing intermediate outputs so that the notebook flows more like a guide.

If you have any questions on this wrt. next steps let me know, I think this is the most important change we'll need to make for this guide to be really useful.